### PR TITLE
HIVE-24383. Add Table type to HPL/SQL (amagyar)

### PR DESCRIPTION
**Pull request by zeroflag**
This patch is for adding the table type support to hplsql as well as BULK COLLECT statements.

### Create a table

A table is an in memory data structure with  hash table like semantics. They can be modeled after a Hive table. There are 3 ways to create a local in memory table:

Let's say we already have a Hive table:

```
CREATE TABLE emp (name string, age int);
insert into emp values('alice', 20);
insert into emp values('bob', 30);
```

To create a table type in hplsql you can do the following:

1. Model it after the emp table:

```
TYPE emp_type IS TABLE OF emp%ROWTYPE INDEX BY BINARY_INTEGER;
```

2. Model it after one column from the emp table:

```
TYPE emp_age_type IS TABLE OF emp.age%TYPE INDEX BY BINARY_INTEGER;
```

3. Define the column type of the table manually:

```
TYPE emp_name_type IS TABLE OF STRING INDEX BY BINARY_INTEGER;
```

In the first case our table is going to have the same columns (name string, age int) as the corresponding hive table (emp) has.

In the second case the table is going to have a single column with the same type as the age column of the emp table. The third case is the same as the second but we define the column type manually (STRING).

Only BINARY_INTEGER index is supported at this moment.

After this we can declare our table as follows:

```
DECLARE rows emp_type;
DECLARE ages emp_age_type;
```

### Populate the table from the DB

The table can be populated by multiple ways.

This puts one row ('alice', 20) into the table at the first index.

```
SELECT * INTO rows(1) FROM emp WHERE name = 'alice';
```

This puts one column of a row into the table at the first index.

```
SELECT age INTO ages(1) FROM emp WHERE name = 'alice';
```

A BULK COLLECT can be used to put multiple rows into the table. This clears the table first and puts everything from the emp Hive table into our local HPLSQL table.

```
SELECT * BULK COLLECT INTO rows FROM emp;
```

BULK COLLECT works with SELECT INTO as well as the FETCH statement.

### Iterate over the table

The NEXT and PRIOR is used to get the subsequent or the previous key of a specific key. Keys are integers, in special cases they are consecutive integers but in general they are like hashtable keys.

```
DECLARE idx INT = rows.FIRST;
WHILE idx IS NOT NULL 
LOOP
  PRINT rows(idx).name || ' = ' || rows(idx).age;
  idx = rows.NEXT(idx);
END LOOP;
```

### Access

A table can be index with a function call syntax.

```
tbl(index)
```

If the table is a ROWTYPE then this returns a ROW where columns can be accessed using the dot syntax.

```
tbl(index).col_name;
```

### Grammar changes

There were multiple problems with the gammer that made harder to implement this data type.

Dots were allowed within an identifier. So foo.bar was a valid variable/procedure name. But this causes ambiguity since package syntax and the dot notation used by the table also use this syntax. The reason this worked before is that packages are second class so func(1,2).some_method() was never allowed with packages but tables work like object they can be returned from a function, and the user might want to chain method calls to the return value.

So the dot being part of the identifier was removed. But in some cases it was kept (as using a different identifier name qident) because fully qualified table names or column names might still contain dots.

1. A new expression type was added for the dot notation syntax. 
2. A new assignment type was added to support this syntax:

```
tbl(index) := 'foo';
```

### Minor features

1. A simple unconditional loop was added.
2. Variable lookup was optimized to use a  hashtable instead of a linear search on an array.

### Implementation notes

A table is stored in a Variable with type = HPL_OBJECT. This works like an object in a OO language. Each object has a class, and the class contains a methodDictionary with all the methods what the object support. The method lookup simply gets the class of an object and find the method in the method dictionary then executes it. There is no inheritance so the lookup always ends here (either it calls the method or fails with an exception if the method was not found). The same class is used for different instances of an object.

Some existing utility packages (UtlFile, DbmOutput) were converted to this mechanism as well.

### CODE REVIEW

File: hplsql/src/main/java/org/apache/hive/hplsql/Select.java [hunk 1]
```diff
@@ -87,37 +94,38 @@ public Integer select(HplsqlParser.Select_stmtContext ctx) {
...
+        } else {
+          trace(ctx, "SELECT INTO statement executed");
+          if (query.next()) {
+            for (int i = 0; i < intoCount; i++) {
+              populateVariable(ctx, query, i);
+            }
+            exec.incRowCount();
+            exec.setSqlSuccess();
+            if (query.next()) {
+              exec.setSqlCode(-1422);
```
[the remark this answers is not in the document]
Comment (zeroflag) [reply hplsql/src/main/java/org/apache/hive/hplsql/Select.java]: I added a new class for that.

File: hplsql/src/main/java/org/apache/hive/hplsql/Select.java [hunk 2]
```diff
@@ -155,7 +162,25 @@ else if (ctx.parent instanceof HplsqlParser.StmtContext) {
...
+        Table table = (Table) var.value;
+        table.populate(query, getIntoTableIndex(ctx, columnIndex), columnIndex);
+      } else if (var.type == Var.Type.ROW) {
+        var.setRowValues(query);
+      } else {
+        var.setValue(query, columnIndex);
+      }
+      exec.trace(ctx, var, query.metadata(), columnIndex);
+    } else {
+      trace(ctx, "Variable not found: " + intoName);
```
[the remark this answers is not in the document]
Comment (zeroflag) [reply hplsql/src/main/java/org/apache/hive/hplsql/Select.java]: I replaced it with an exception.

File: hplsql/src/main/java/org/apache/hive/hplsql/Stmt.java [hunk 3]
```diff
@@ -961,6 +987,22 @@ public Integer forRange(HplsqlParser.For_range_stmtContext ctx) {
...
+    trace(ctx, "UNCONDITIONAL LOOP - ENTERED");
+    String label = exec.labelPop();
+    while (true) {
+      exec.enterScope(Scope.Type.LOOP);
+      visit(ctx.block());
+      exec.leaveScope();
+      if (canContinue(label)) {
+        continue;
+      }
+      break;
```
[the remark this answers is not in the document]
Comment (zeroflag) [reply hplsql/src/main/java/org/apache/hive/hplsql/Stmt.java]: Ok.

File: hplsql/src/main/java/org/apache/hive/hplsql/Var.java [hunk 4]
```diff
@@ -26,13 +26,14 @@
 
 import org.apache.hive.hplsql.executor.QueryResult;
 
+
```
[the remark this answers is not in the document]
Comment (zeroflag) [reply hplsql/src/main/java/org/apache/hive/hplsql/Var.java]: removed.

File: hplsql/src/main/java/org/apache/hive/hplsql/objects/HplClass.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.apache.hive.hplsql.objects;
+
+public interface HplClass {
+  HplObject instantiate();
```
[the remark this answers is not in the document]
Comment (zeroflag) [reply hplsql/src/main/java/org/apache/hive/hplsql/objects/HplClass.java]: That makes more sense. I renamed it.

**Review by mustafaiman — APPROVED** [review]

**Pull request merged**